sensor/DeviceServiceScanSensor.js

```diff
@@ -15,138 +15,250 @@
 'use strict';
 
 const log = require('../net2/logger.js')(__filename);
-
-const util = require('util');
-
 const Sensor = require('./Sensor.js').Sensor;
-
+const extensionManager = require('./ExtensionManager.js')
+const fs = require('fs');
+const Promise = require('bluebird');
+Promise.promisifyAll(fs);
+const featureName = "device_service_scan";
+const policyKeyName = "device_service_scan";
+const HostManager = require("../net2/HostManager.js");
+const hostManager = new HostManager();
 const rclient = require('../util/redis_manager.js').getRedisClient();
-
+const sysManager = require('../net2/SysManager.js');
 const cp = require('child_process');
-
+const util = require('util');
 const Firewalla = require('../net2/Firewalla');
-
 const xml2jsonBinary = Firewalla.getFirewallaHome() + "/extension/xml2json/xml2json." + Firewalla.getPlatform();
 
-const fc = require('../net2/config.js');
-
-const HostManager = require("../net2/HostManager.js");
-const hostManager = new HostManager();
-
-const sysManager = require('../net2/SysManager.js');
-const sem = require('../sensor/SensorEventManager.js').getInstance();
-
 class DeviceServiceScanSensor extends Sensor {
-  constructor() {
-    super();
-  }
-
-  run() {
-    let firstScanTime = this.config.firstScan * 1000 || 120 * 1000; // default to 120 seconds
-    setTimeout(async() => {
-      await this.checkAndRunOnce();
-      sem.emitEvent({
-        type: "DeviceServiceScanComplete",
-        message: ""
-      });
-    }, firstScanTime);
+    async run() {
+        this.systemSwitch = false;
+        this.adminSystemSwitch = false;
+        this.deviceScanMap = {};
+        this.tagScanMap = {};
+        this.networkScanMap = {};
+        this.systemScanTimer = '';
+        this.interval = this.config.interval * 1000 || 30 * 60 * 1000; // 30 minutes
+        extensionManager.registerExtension(policyKeyName, this, {
+            applyPolicy: this.applyPolicy
+        });
+        this.hookFeature(featureName);
+    }
 
-    let interval = this.config.interval * 1000 || 30 * 60 * 1000; // 30 minutes
-    setInterval(() => {
-      this.checkAndRunOnce();
-    }, interval);
-  }
+    async job() {
+        await this.applyScan();
+    }
 
-  isSensorEnable() {
-    return fc.isFeatureOn("device_service_scan");
-  }
+    async apiRun() {
+    }
 
-  async checkAndRunOnce() {
-    try {
-      let result = await this.isSensorEnable();
-      if (result) {
-        return await this.runOnce();
-      };
-    } catch(err) {
-      log.error('Failed to scan: ', err);
-    };
+    async applyPolicy(host, ip, policy) {
+        log.info("Applying device service scan policy:", ip, policy);
+        try {
+            if (ip === '0.0.0.0') {
+                if (policy === true) {
+                    this.systemSwitch = true;
```

Comment:
looks like `systemSwitch` and `adminSystemSwitch` are the same thing.
policy:system all targets 0.0.0.0

Reply:
Nope, they aer different, adminSystemSwitch means feature on or off,  systemSwitch means target=>0.0.0.0 policy is true or  false.

Reply:
consider using `Config.isFeatureOn()` maybe? this is confusing

Reply:
consider using `Config.isFeatureOn()` maybe? this is confusing

Reply:
`Config.isFeatureOn()` is fine, but this sensor inherits `senor`, will call `hookFeature` which may listenning feature on and off, so we apply a new variable to handle this. Just keep the structure currently.